File: frontend/src/services/reportService.js
/**
 * @author  ThangNB - HE201024
 * @created 2026-07-19
 * @updated 2026-07-19
 *
 * API client for the Clinic Manager's reports — UC-49 (operational
 * dashboard), UC-50 (revenue), UC-51 (patient statistics), UC-52 (staff
 * performance) and UC-53 (feedback), plus their CSV exports.
 *
 * All endpoints are read-only; access is restricted to MANAGER / ADMIN
 * server-side.
 */
import axiosClient from '../api/axiosClient'

/**
 * Builds the axios params for a date range, omitting empty bounds so the
 * backend applies its own default (first of this month → today).
 *
 * @param {string} [from] ISO yyyy-MM-dd start
 * @param {string} [to]   ISO yyyy-MM-dd end
 * @returns {{params: Object}} axios config fragment
 */
const range = (from, to) => {
  const p = {}
  if (from) p.from = from
  if (to) p.to = to
  return { params: p }
}

export const reportService = {
  /** UC-49: today's live operational figures. @returns {Promise} dashboard widgets */
  operationalDashboard: () => axiosClient.get('/v1/reports/dashboard'),

  /** UC-50: revenue by service, doctor and payment method.
   *  @param {string} [from] @param {string} [to] @returns {Promise} */
  revenue: (from, to) => axiosClient.get('/v1/reports/revenue', range(from, to)),

  /** UC-51: patient volume, new vs returning, top diagnoses.
   *  @param {string} [from] @param {string} [to] @returns {Promise} */
  patientStatistics: (from, to) => axiosClient.get('/v1/reports/patient-statistics', range(from, to)),

  /** UC-52: per-doctor KPIs.
   *  @param {string} [from] @param {string} [to] @returns {Promise} */
  staffPerformance: (from, to) => axiosClient.get('/v1/reports/staff-performance', range(from, to)),

  /** UC-53: aggregated patient feedback.
   *  @param {string} [from] @param {string} [to] @returns {Promise} */
  feedbackReport: (from, to) => axiosClient.get('/v1/reports/feedback', range(from, to)),

  // ── Exports, resolving to a Blob ──
  // Note: the backend emits UTF-8 CSV, whereas UC-50 step 6 specifies .xlsx.

  /** UC-50: exports the revenue report. @returns {Promise<Blob>} */
  exportRevenue: (from, to) =>
    axiosClient.get('/v1/reports/revenue/export', { ...range(from, to), responseType: 'blob' }),
  /** UC-51: exports the patient statistics. @returns {Promise<Blob>} */
  exportPatientStatistics: (from, to) =>
    axiosClient.get('/v1/reports/patient-statistics/export', { ...range(from, to), responseType: 'blob' }),

  /** UC-53: exports the feedback report. @returns {Promise<Blob>} */
  exportFeedback: (from, to) =>
    axiosClient.get('/v1/reports/feedback/export', { ...range(from, to), responseType: 'blob' }),
}

/**
 * Triggers a browser download for a Blob returned by one of the export calls.
 *
 * The object URL is revoked on a delay rather than immediately, because
 * revoking it before the browser has started the download cancels the save in
 * some browsers.
 *
 * @param {Blob} blob     file contents
 * @param {string} filename name offered to the user
 */
export function downloadBlob(blob, filename) {
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  document.body.removeChild(a)
  setTimeout(() => URL.revokeObjectURL(url), 60000)
}

File: backend/src/main/java/com/ecms/service/ReportService.java
package com.ecms.service;

import jakarta.servlet.http.HttpServletResponse;

import java.io.IOException;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

/**
 * @author      ThangNB - HE201024
 * @contributor Đồng Mạnh Hùng - HE200743
 * @created     2026-05-31
 * @updated     2026-07-19
 *
 * Reporting and analytics contract for the Clinic Manager
 * (UC-49, UC-50, UC-51, UC-52, UC-53).
 *
 * Results are returned as {@code Map} / {@code List<Map>} rather than fixed
 * DTOs so the frontend can render tables and charts off the same payload
 * without a new type per widget.
 *
 * Every method is read-only and enforces no business rule; access is gated by
 * role in SecurityConfig.
 */
public interface ReportService {

    /**
     * UC-49: live operational figures for today — appointment progress, queue
     * length per doctor, pending prescriptions, outstanding invoices and lab
     * orders in progress.
     *
     * @return dashboard widgets keyed by metric name
     */
    Map<String, Object> operationalDashboard();

    /**
     * UC-50: revenue over a period, broken down by service category, doctor
     * and payment method.
     *
     * Counts PAID invoices by {@code paidAt}, so revenue lands on the day the
     * money arrived rather than the day the invoice was raised.
     *
     * @param from period start, inclusive
     * @param to   period end, inclusive
     * @return totals plus the per-dimension breakdowns
     */
    Map<String, Object> revenueReport(LocalDate from, LocalDate to);

    /**
     * UC-51: patient volume over a period — visits, new vs returning patients,
     * top diagnoses and appointment status distribution.
     *
     * @param from period start, inclusive
     * @param to   period end, inclusive
     * @return statistics keyed by metric name
     */
    Map<String, Object> patientStatistics(LocalDate from, LocalDate to);

    /**
     * UC-52: per-doctor KPIs over a period — patients seen, average
     * consultation duration and prescription volume.
     *
     * @param from period start, inclusive
     * @param to   period end, inclusive
     * @return one row per staff member
     */
    List<Map<String, Object>> staffPerformance(LocalDate from, LocalDate to);

    /**
     * UC-53: aggregated patient feedback over a period — average rating per
     * doctor, response counts and common themes.
     *
     * @param from period start, inclusive
     * @param to   period end, inclusive
     * @return feedback analytics keyed by metric name
     */
    Map<String, Object> feedbackReport(LocalDate from, LocalDate to);

    // ── Exports ───────────────────────────────────────────────────────────────
    // Emitted as UTF-8 CSV, which Excel opens directly. Note this is a
    // deviation from UC-50 step 6, which specifies a real .xlsx file.

    /**
     * UC-50 step 5-6: exports the revenue report.
     *
     * @param from     period start, inclusive
     * @param to       period end, inclusive
     * @param response servlet response the CSV is streamed to
     * @throws IOException if the response stream fails
     */
    void exportRevenueCsv(LocalDate from, LocalDate to, HttpServletResponse response) throws IOException;

    /**
     * UC-51 step 4: exports the patient statistics.
     *
     * @param from     period start, inclusive
     * @param to       period end, inclusive
     * @param response servlet response the CSV is streamed to
     * @throws IOException if the response stream fails
     */
    void exportPatientStatisticsCsv(LocalDate from, LocalDate to, HttpServletResponse response) throws IOException;

    /**
     * UC-53 step 5: exports the detailed feedback list.
     *
     * @param from     period start, inclusive
     * @param to       period end, inclusive
     * @param response servlet response the CSV is streamed to
     * @throws IOException if the response stream fails
     */
    void exportFeedbackCsv(LocalDate from, LocalDate to, HttpServletResponse response) throws IOException;
}
